core/capabilities/vault/gw_handler.go

```diff
@@ -0,0 +1,312 @@
+
+	HandlerName = "VaultHandler"
+)
+
+type metrics struct {
+	// Given that all requests are coming from the gateway, we can assume that all errors are internal errors
+	requestInternalError metric.Int64Counter
+	requestSuccess       metric.Int64Counter
+}
+
+func newMetrics() (*metrics, error) {
+	requestInternalError, err := beholder.GetMeter().Int64Counter("vault_node_request_internal_error")
+	if err != nil {
+		return nil, fmt.Errorf("failed to register internal error counter: %w", err)
+	}
+
+	requestSuccess, err := beholder.GetMeter().Int64Counter("vault_node_request_success")
+	if err != nil {
+		return nil, fmt.Errorf("failed to register success counter: %w", err)
+	}
+
+	return &metrics{
+		requestInternalError: requestInternalError,
+		requestSuccess:       requestSuccess,
+	}, nil
+}
+
+type GatewayHandler struct {
+	capRegistry    core.CapabilitiesRegistry
+	secretsService SecretsService
+	gwConnector    core.GatewayConnector
+	lggr           logger.Logger
+	metrics        *metrics
+}
+
+func NewGatewayHandler(capabilitiesRegistry core.CapabilitiesRegistry, secretsService SecretsService, gwConnector core.GatewayConnector, lggr logger.Logger) (*GatewayHandler, error) {
+	metrics, err := newMetrics()
+	if err != nil {
+		return nil, fmt.Errorf("failed to create metrics: %w", err)
+	}
+
+	return &GatewayHandler{
+		capRegistry:    capabilitiesRegistry,
+		secretsService: secretsService,
+		gwConnector:    gwConnector,
+		lggr:           lggr.Named(HandlerName),
+		metrics:        metrics,
+	}, nil
+}
+
+func (h *GatewayHandler) Start(ctx context.Context) error {
+	return nil
+}
+
+func (h *GatewayHandler) Close() error {
+	return nil
+}
+
+func (h *GatewayHandler) ID(ctx context.Context) (string, error) {
+	return HandlerName, nil
+}
+
+func (h *GatewayHandler) HandleGatewayMessage(ctx context.Context, gatewayID string, req *jsonrpc.Request[json.RawMessage]) (err error) {
+	h.lggr.Debugf("Received message from gateway %s: %v", gatewayID, req)
+
+	var response *jsonrpc.Response[json.RawMessage]
+	switch req.Method {
+	case vault_api.MethodSecretsCreate:
+		response = h.handleSecretsCreate(ctx, gatewayID, req)
+	case vault_api.MethodSecretsGet:
+		response = h.handleSecretsGet(ctx, gatewayID, req)
+	default:
+		response = h.errorResponse(ctx, gatewayID, req, api.UnsupportedMethodError, errors.New("unsupported method: "+req.Method))
+	}
+
+	if err = h.gwConnector.SendToGateway(ctx, gatewayID, response); err != nil {
+		h.lggr.Errorf("Failed to send message to gateway %s: %v", gatewayID, err)
+		return err
+	}
+
+	h.lggr.Infof("Sent message to gateway %s: %v", gatewayID, response)
+	h.metrics.requestSuccess.Add(ctx, 1, metric.WithAttributes(
+		attribute.String("gateway_id", gatewayID),
+	))
+	return nil
+}
+
+func (h *GatewayHandler) handleSecretsCreate(ctx context.Context, gatewayID string, req *jsonrpc.Request[json.RawMessage]) *jsonrpc.Response[json.RawMessage] {
+	var requestData vault_api.SecretsCreateRequest
```

Comment:
We should make this a batch req/response; same with the Get method

Reply:
Yes, will do all in a follow-up PR